Revamping lifecycle docs (#75)
r-lib/lifecycle: vignettes/user.Rmd

---
title: "Scripts"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Scripts}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
options("lifecycle:::calling_package" = "tibble")
```

The lifecycle package uses warnings to advertise which functions are deprecated.
Deprecated functions will be removed in a future release, so it's good practice to eliminate the warnings as soon as you see them.

```{r setup}
library(lifecycle)
```

For example, lets imagine your code used `tibble::data_frame()`, which was deprecated in favour of `tibble()` in version 1.1.0.
`data_frame()` now looks something like this:

```{r}
data_frame <- function(...) {
  deprecate_warn("1.1.0", "data_frame()", "tibble()")
  tibble::tibble(...)
}
```

So if you use `data_frame()` in your own code you'll get a warning:

> ```{r}

Reviewer: Hard-code the output in prevision of `data_frame()` becoming defunct?

Author: I defined my own `data_frame()` above 😄

df1 <- data_frame(x = 1, y = 2)
df2 <- data_frame(a = "apple", b = "banana")
```

> You'll notice that the warning only appears the first time we call it --- lifecycle only notifies you every 8 hours to avoid driving you mad if you've used a deprecated function in many places.

Reviewer: ```suggestion
You'll notice that the warning only appears the first time we call it --- lifecycle only notifies you every 8 hours to avoid disrupting your work if you've used a deprecated function in many places.
```

So how do you track down exactly where the warning came from?

> lifecycle provides too main tools.

Reviewer: ```suggestion
lifecycle provides two main tools.
```

Firstly, if you're ready to spend some time tracking down deprecated functions, set the `lifecycle_verbosity` option.
You set it to warning to make it warn every time:

```{r}
options("lifecycle_verbosity" = "warning")
df1 <- data_frame(x = 1, y = 2)
df2 <- data_frame(a = "apple", b = "banana")
```

When using warnings, you can track down the source with `lifecycle::last_warning()`:

```{r, eval = FALSE}
lifecycle::last_warning()
#> <deprecated>
#> message: `data_frame()` was deprecated in tibble 1.1.0.
#> Please use `tibble()` instead.
#> Backtrace:
#>  1. tibble::data_frame(x = 1, y = 2)
```

This provides a back trace that shows you the full sequence of calls that lead to the deprecated function.

> 

Reviewer: ```suggestion
The plural variant `lifecycle::last_warnings()` shows backtraces for all the warnings that were signalled in the last command. If you're a package developer this is especially helpful in conjunction with `devtools::test()`.

```
